linker.go

```diff
@@ -111,50 +111,100 @@ func findReferences(progs map[string]*ProgramSpec) error {
 	return nil
 }
 
-// collectCORERelos returns a list of CO-RE relocations of the layout's progs
-// in order.
-func collectCORERelos(layout []reference) (btf.CORERelos, error) {
-	var relos btf.CORERelos
-	for _, sym := range layout {
-		if sym.spec.BTF == nil {
-			return nil, fmt.Errorf("program %s: missing BTF", sym.spec.Name)
+type funcInfoMeta struct{}
+type coreRelocationMeta struct{}
+
+// Assign per-section metadata from BTF to a section's instructions.
+func assignExtInfos(insns asm.Instructions, ext *btf.ExtInfos, section string) error {
+	funcInfos := ext.FuncInfos[section]
+	lineInfos := ext.LineInfos[section]
+	reloInfos := ext.RelocationInfos[section]
+
+	iter := insns.Iterate()
+	for iter.Next() {
+		if len(funcInfos) > 0 && funcInfos[0].Offset == iter.Offset {
+			iter.Ins.Metadata.Set(funcInfoMeta{}, funcInfos[0].Func)
+			funcInfos = funcInfos[1:]
+		}
+
+		if len(lineInfos) > 0 && lineInfos[0].Offset == iter.Offset {
+			*iter.Ins = iter.Ins.WithSource(lineInfos[0].Line)
+			lineInfos = lineInfos[1:]
+		}
+
+		if len(reloInfos) > 0 && reloInfos[0].Offset == iter.Offset {
+			iter.Ins.Metadata.Set(coreRelocationMeta{}, reloInfos[0].Relocation)
+			reloInfos = reloInfos[1:]
 		}
-		relos = append(relos, sym.spec.BTF.CORERelos.Offset(uint32(sym.offset))...)
 	}
 
-	return relos, nil
+	return nil
 }
 
-// marshalFuncInfos returns the BTF func infos of all progs in order.
-func marshalFuncInfos(layout []reference) ([]byte, error) {
-	if len(layout) == 0 {
-		return nil, nil
+// marshalExtInfos encodes function and line info embedded in insns into kernel
+// wire format.
+func marshalExtInfos(insns asm.Instructions) (funcInfos, lineInfos []byte, _ error) {
```

Comment:
Could this live in package btf, or is there an issue with a dependency on `asm`?

My hope was to get rid of FuncInfo and LineInfo, since all the information they carry can now be represented by `asm.Instructions`. e.g. *Func and the instruction's position in the slice as you've shown below.

Ultimately, we should end up with only Line and Func, and their respective wire formats, without the `Func/LineInfo` intermediates.